notewhacker.lisp

```diff
@@ -407,18 +417,26 @@ difficulty. Return the new target."
       ;; TODO: Better choice at random notes. Learning? (Make player try
       ;; and hit difficult notes, which player had problems with
       ;; earlier.)
-      (let ((base-key (if (eq (clef staff) 'f-clef)
-                          30
-                          55
-                          )))
+      (let* ((base-key (min-base-key staff))
+             (staff-width (- (max-base-key staff) base-key))
+             (max-chord-width 12) ; octave
+             (half-chord-width (/ max-chord-width 2)))
         ;; NOTE: At the moment, this is tuned for the keyboard input.
-        (flet ((random-note ()
-                 (+ base-key (random 17))
-                 ))
+        (labels ((chord-base-note (first-note)
+                  (let ((delta-min (- first-note base-key))
```

Comment:
This is a bit hard to read (labels can also have docstrings), and using the `half-chord-width` near the ends of the range, combined with a perhaps odd value of `max-chord-width`, is a bit problematic regarding the uniformity of the random integers received.

My suggestion is to create a couple of helper functions (in top level or with `labels`) like this, and use it in the `let` below.

```
(defun pick-random (l r width)
  "Pick a random integer from range [R, L] (inclusive).

Return three values: the random integer, and new, tighter left
and right limits that are both within at maximum WIDTH distance
from the chosen random integer."
  (let* ((rnd-val (+ l (random (- r l))))
         (new-l (max l (- rnd-val width)))
         (new-r (min r (+ rnd-val width))))
    (values rnd-val new-l new-r)))

(defun pick-max-n-randoms (num l-limit r-limit width)
  "Pick at maximum NUM unique random integers, maybe less, from a
range of integers [L-LIMIT, R-LIMIT] (inclusive) so that all of
the random integers are within WIDTH of each other - that is, the
returned list's max - min < WIDTH. Return a list of these random
integers.

Note that random is called at max NUM times - no guarantee is
made that NUM integers will be returned. The smaller WIDTH/NUM
is, the less probable it is to get all NUM values.

Returned integers are in random order and unique."
  (let (randoms)
    (loop :repeat num
          :do (multiple-value-bind (v new-l new-r) (pick-random l-limit r-limit width)
                (setf l-limit new-l
                      r-limit new-r)
                (pushnew v randoms))
          :finally (return randoms))))
```

And then the
```(let ((note-num ...```

could be written like
```
          (let* ((note-num (1+ (random (ceiling (/ (1+ score) 500)))))
                ((pick-max-n-randoms note-num (min-base-key staff) (max-base-key staff) max-chord-width notes)))
```